internal/chain/engine.go

```diff
@@ -324,6 +329,44 @@ func (e *engineImpl) Finalize(
 			}
 		}
 	}
+
+	if inStakingEra {
+		if isBeaconChain {
+			superCommittee, err := chain.ReadShardState(chain.CurrentHeader().Epoch())
+
+			if err != nil {
+				return nil, nil, err
+			}
+			processed := make(map[common.Address]struct{})
+			for i := range superCommittee.Shards {
+				shard := superCommittee.Shards[i]
+				for j := range shard.Slots {
+					slot := shard.Slots[j]
+					if slot.EffectiveStake != nil { // For external validator
+						_, ok := processed[slot.EcdsaAddress]
+						if !ok {
+							processed[slot.EcdsaAddress] = struct{}{}
+						}
+					}
+				}
+			}
+
+			if err := availability.IncrementValidatorSigningCounts(
+				chain, header, header.ShardID(), state, processed,
+			); err != nil {
+				return nil, nil, err
+			}
+			if err := availability.SetInactiveUnavailableValidators(
```

Comment:
And with this, we should put the tallying before the shard state proposal, so kicked out nodes won't be part of shard anymore. So this part should be separate from IncrementValidatorSigningCounts.